zebra-chain/src/orchard/shielded_data.rs

```diff
@@ -48,6 +50,37 @@ impl ShieldedData {
     pub fn nullifiers(&self) -> impl Iterator<Item = &Nullifier> {
         self.actions().map(|action| &action.nullifier)
     }
+
+    /// Calculate the Action binding verification key.
+    ///
+    /// Getting the binding signature validating key from the Spend and Output
+    /// description value commitments and the balancing value implicitly checks
+    /// that the balancing value is consistent with the value transfered in the
+    /// Spend and Output descriptions but also proves that the signer knew the
+    /// randomness used for the Spend and Output value commitments, which
+    /// prevents replays of Output descriptions.
+    ///
+    /// The net value of Spend transfers minus Output transfers in a transaction
+    /// is called the balancing value, measured in zatoshi as a signed integer
+    /// v_balance.
+    ///
+    /// Consistency of v_balance with the value commitments in Spend
+    /// descriptions and Output descriptions is enforced by the binding
+    /// signature.
+    ///
+    /// Instead of generating a key pair at random, we generate it as a function
+    /// of the value commitments in the Spend descriptions and Output
+    /// descriptions of the transaction, and the balancing value.
+    ///
+    /// https://zips.z.cash/protocol/protocol.pdf#orchardbalance
+    pub fn binding_verification_key(&self) -> redpallas::VerificationKeyBytes<Binding> {
+        let cv: ValueCommitment = self.actions().map(|action| action.cv).sum();
+        let cv_balance: ValueCommitment =
+            ValueCommitment::new(pallas::Scalar::zero(), self.value_balance);
+
+        let key_bytes: [u8; 32] = (cv - cv_balance).into();
```

Comment:
We only have one `cv` field in orchard so i think this is correct but not sure.

Reply:
Zebra's Sapling value balance code is probably wrong, see bug #1939.

We should implement this derivation based on the Orchard consensus rules:
https://zips.z.cash/protocol/protocol.pdf#orchardbalance

Reply:
We found the proximate source of the bug in our implementation of sighash, so I feel confident that this value balance code is actually correct now 👍

Reply:
Yep #1939 is fixed.